common/flatpak-transaction.c

```diff
@@ -4062,6 +4092,272 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+          dir_ref_is_installed (priv->dir, op->ref, NULL, &deploy_data))
+        {
+          g_autofree char *full_previous_runtime = NULL;
+          g_autofree char *prev_runtime_remote = NULL;
+
+          const char *previous_runtime = flatpak_deploy_data_get_runtime (deploy_data);
+
+          if (previous_runtime == NULL || *previous_runtime == '\0')
+            continue;
+
+          full_previous_runtime = g_strconcat ("runtime/", previous_runtime, NULL);
+          if (g_strcmp0 (full_previous_runtime, runtime_ref) == 0)
+            continue;
+
+          g_clear_pointer (&related, g_ptr_array_unref);
+          related = find_related_from_deploy (self, full_previous_runtime, &prev_runtime_remote);
+
+          g_hash_table_replace (maybe_unused_runtimes,
+                                g_strdup (full_previous_runtime),
+                                g_strdup (prev_runtime_remote));
+
+          g_ptr_array_add (to_be_excluded, op->ref);
+
+          g_ptr_array_add (run_after_ops, op);
+
+          if (priv->disable_related)
+            continue;
+
+          for (i = 0; related && i < related->len; i++)
+            {
+              FlatpakRelated *rel = g_ptr_array_index (related, i);
+              if (rel->delete)
+                g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (prev_runtime_remote));
+            }
+        }
+    }
+
+  /* Subtract newly_used_runtimes from maybe_unused_runtimes */
+  GLNX_HASH_TABLE_FOREACH (newly_used_runtimes, const char *, runtime_ref)
+    g_hash_table_remove (maybe_unused_runtimes, runtime_ref);
+
+  if (g_hash_table_size (maybe_unused_runtimes) == 0)
+    return TRUE;
+
+  /* Check which things in maybe_unused_runtimes will be unused after the
+   * uninstall ops in the transaction are executed. Note that
+   * maybe_unused_runtimes and to_be_excluded are modified in this helper */
+  if (!prune_maybe_unused_list (self, maybe_unused_runtimes, to_be_excluded,
+                                cancellable, error))
+    return FALSE;
+
+  /* Schedule each unused runtime to be uninstalled */
+  GLNX_HASH_TABLE_FOREACH_KV (maybe_unused_runtimes, const char *, runtime_ref, const char *, remote)
+    {
+      g_autofree char *resolved_remote = NULL;
+      FlatpakTransactionOperation *unused_op = NULL;
+
+      if (remote == NULL)
+        g_assert (dir_ref_is_installed (priv->dir, runtime_ref, &resolved_remote, NULL));
+      else
+        resolved_remote = g_strdup (remote);
+
+      unused_op = flatpak_transaction_add_op (self, resolved_remote, runtime_ref,
+                                              NULL, NULL, NULL, NULL,
+                                              FLATPAK_TRANSACTION_OPERATION_UNINSTALL);
+
+      for (i = 0; i < run_after_ops->len; i++)
+        {
+          FlatpakTransactionOperation *op = g_ptr_array_index (run_after_ops, i);
+          run_operation_before (op, unused_op, 1);
+        }
+    }
+
```

Comment:
I wonder if we could do better here both in terms for performance as well as code duplication if we merge all the handling of uninstall related refs dependency to one place.

 I.e. first we iterate over all the ops like this code does, but in the "app was uninstalled, and it had a runtime that may be unused now" as well as the "app was updated to new runtime, and the old runtime may be unused now" we just add the runtime ref to a maybe_unused_runtimes_without_deps hash. 

And then after all the ops are handled we iterate over without_deps and find related refs that also may be unused and add them all to maybe_unused_runtimes. That way we only do the computation once per repeated runtime, and additionally we don't have to duplicate the code in both cases.

Reply:
As I said [above](https://github.com/flatpak/flatpak/pull/3783#discussion_r490627762) I think we do also have to account for install operations and their dependencies.

Reply:
And we also have to account for related refs of the op ref not just of its runtime, as they might become unused if the op is an update.

A lot of the complexity here comes from the desire to only add uninstall ops for things that are made unused by something in the transaction when `priv->include_unused_uninstall_ops == FALSE`. If we wanted to just uninstall the whole set of runtimes that will be considered unused after the transaction, for every transaction, that would be a lot easier. But then you would sometimes run into a situation where a transaction should only be affecting one thing A and it tries to uninstall something unrelated B, if for example the previous transaction which made B unused had `priv->disable_deps == TRUE`, or in the other situations I mentioned in the above thread.